File: src/Canon.Emit/OpenSearchEmitter.fs
namespace Canon.Emit.OpenSearch

open Canon.Emit
open Canon.Introspect

/// Emits an OpenSearch/Elasticsearch index mapping derived from the domain schema.
/// Converts typed fields into strict OpenSearch field types.
type OpenSearchEmitter() =
    
    // Map standard data types to OpenSearch data types.
    let mapDataType (sqlType: string) =
        match sqlType.ToLowerInvariant() with
        | "integer" | "int" -> "integer"
        | "bigint" -> "long"
        | "boolean" -> "boolean"
        | "timestamp" | "date" -> "date"
        | "decimal" | "numeric" -> "double" // Simplified
        | _ -> "keyword" // Default to keyword for strictness (instead of text)
        
    interface IEmitter with
        member this.Emit(tables: TableDef list) =
            tables |> List.map (fun table ->
                let props = 
                    table.Columns 
                    |> List.map (fun col ->
                        $"""        "{col.Name}": {{ "type": "{mapDataType col.DataType}" }}"""
                    )
                    |> String.concat ",\n"
                
                let mapping = $"""{{
  "mappings": {{
    "properties": {{
{props}
    }}
  }}
}}"""
                mapping, Canon.Core.Fidelity.Approximate "OpenSearch drops foreign keys, constraints, and defaults"
            )

File: playground/src/canonflow/Canon.Emit/OpenSearchEmitter.js

import { map } from "../fable_modules/fable-library-js.5.6.0/List.js";
import { join, concat } from "../fable_modules/fable-library-js.5.6.0/String.js";
import { Fidelity, Divergence } from "../Canon.Core/Lineage.js";
import { Emitter } from "./Emitter.js";

export function mapDataType(sqlType) {
    const matchValue = sqlType.toLowerCase();
    switch (matchValue) {
        case "integer":
        case "int":
            return "integer";
        case "bigint":
            return "long";
        case "boolean":
            return "boolean";
        case "timestamp":
        case "date":
            return "date";
        case "decimal":
        case "numeric":
            return "double";
        default:
            return "keyword";
    }
}

export function createEmitter() {
    return new Emitter((tables) => map((table) => [concat("{\n  \"mappings\": {\n    \"properties\": {\n", join(",\n", map((col) => (`        "${col.Name}": { "type": "${mapDataType(col.DataType)}" }`), table.Columns)), "\n    }\n  }\n}"), new Fidelity(2, [new Divergence(1, ["OpenSearch drops foreign keys, constraints, and defaults"])])], tables));
}

